### ISSUE TITLE
Binary compatibility issues on Python 3.7/Linux

### ISSUE THREAD

**Issue by penguinolog** [gevent/gevent#1260]
* gevent version: 1.3.5 (from PyPI)
* Python version: cPython 3.7.0
* Operating System: Gentoo Linux, Python 3.7 docker image

### Description:

Run at python console: `import gevent.threadpool`

### Warnings:

At virtualenv@gentoo linux
```
/home/penguinolog/virtualenvs/py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/home/penguinolog/virtualenvs/py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/home/penguinolog/virtualenvs/py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/home/penguinolog/virtualenvs/py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/home/penguinolog/virtualenvs/py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/home/penguinolog/virtualenvs/py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/home/penguinolog/virtualenvs/py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/home/penguinolog/virtualenvs/py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
```

At python:3.7 docker image:
```
/usr/local/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/usr/local/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/usr/local/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/usr/local/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/usr/local/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/usr/local/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/usr/local/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
/usr/local/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144, got 128
  return f(*args, **kwds)
```

### Result:
```
>>> executor = gevent.threadpool.ThreadPool(maxsize=20,hub=None)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/home/penguinolog/virtualenvs/py37/lib/python3.7/site-packages/gevent/threadpool.py", line 63, in __init__
    hub = get_hub()
  File "src/gevent/_hub_local.py", line 73, in gevent.__hub_local.get_hub_noargs
  File "src/gevent/_hub_local.py", line 80, in gevent.__hub_local.get_hub_noargs
  File "/home/penguinolog/virtualenvs/py37/lib/python3.7/site-packages/gevent/hub.py", line 390, in __init__
    WaitOperationsGreenlet.__init__(self, None, None)
  File "src/gevent/_greenlet_primitives.py", line 42, in gevent.__greenlet_primitives.TrackedRawGreenlet.__init__
SystemError: Objects/dictobject.c:1439: bad argument to internal function
```
```
>>> executor = gevent.threadpool.ThreadPool(maxsize=20,hub=None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/gevent/threadpool.py", line 63, in __init__
    hub = get_hub()
  File "src/gevent/_hub_local.py", line 73, in gevent.__hub_local.get_hub_noargs
  File "src/gevent/_hub_local.py", line 80, in gevent.__hub_local.get_hub_noargs
  File "/usr/local/lib/python3.7/site-packages/gevent/hub.py", line 390, in __init__
    WaitOperationsGreenlet.__init__(self, None, None)
  File "src/gevent/_greenlet_primitives.py", line 42, in gevent.__greenlet_primitives.TrackedRawGreenlet.__init__
SystemError: Objects/dictobject.c:1439: bad argument to internal function
```

Temporary solution:
` pip install -U --force-reinstall --no-binary :all: gevent`
(Compile gevent during installation)

**Comment by penguinolog** [comment]
De-facto required to re-build gevent only, so `:all:` rule is overkill

**Comment by jamadden** [comment]
What version of greenlet is installed?

**Comment by penguinolog** [comment]
```
root@34710e4bd3c4:/# pip freeze
gevent==1.3.5
greenlet==0.4.14
```

**Comment by penguinolog** [comment]
Issue only if wheel package installed from PyPI

**Comment by jamadden** [comment]
Ok, gevent wheels are built with greenlet 0.4.13. Something must have changed in the ABI of greenlet 0.4.14 (which was released last night).

As you discovered, building gevent from source should fix the issue, and I would expect installing greenlet==0.4.13 would as well.

**Comment by pkittenis** [comment]
Greenlet did make an [ABI change](https://github.com/python-greenlet/greenlet/commit/8d998cb79ebea348a8f5258b642a1f772ecd93b8) to handle a [Python 3.7 crash](https://github.com/python-greenlet/greenlet/issues/131).

The binary incompatibility warning from greenlet [has been raised separately](https://github.com/python-greenlet/greenlet/issues/135).

**Comment by jamadden** [comment]
Wheels cannot be rebuilt without changing the version number. And gevent has to include its own copy of `greenlet.h` to support certain installation scenarios using sdists so that also requires a new version number.

**Comment by pkittenis** [comment]
Yes, meant with version change. Thanks for making the changes :+1:

**Comment by AndreyBulezyuk** [comment]
I had the same issue:
- Running Docker Python3.7 on Google Cloud Kubernetes.
- Gunicorn, Greenlet and Gevent newest versions

Issues: Workers were timing out, after some fixing, the Runtime Error.

Fixed:
- Changed Base Image to Python3.6
- gunicorn==19.9.0
- gevent==1.3.4
- greenlet==0.4.13

Thanks for your help!

**Comment by JingjingLei** [comment]
when i run dokcer python server with gunicorn  i got this problem,  i think it's my config.py is incorrect at first, i tried every way, but problem  still unsolved,  that really make me creay. I didn't expect the version to be a problem. thank you very much

**Comment by weakish** [comment]
It's all history repeating itself. greenlet releasesd yesterday (2020-09-22) brings another ABI breaking change:

```
RuntimeWarning: greenlet.greenlet size changed, may indicate binary incompatibility.
Expected 144 from C header, got 152 from PyObject ...
```

Tested under Ubuntu 20.04 and gevent 1.5.0

**Comment by jamadden** [comment]
Please do not add unrelated comments to closed issues. Open a new issue if you have, well, a new issue.

gevent 20.9.0 was released at the same time greenlet 0.4.17 was un-yanked. On Python 3.7 and higher, gevent 20.9.0 is required to use (a standard build of) greenlet 0.4.17.

gevent 1.5 doesn't support Python 3.8.
